### ISSUE TITLE
exclusions are install dependent and documented example for terraform_valudate is incorrect 

### ISSUE THREAD

**Issue by Noel-Jones** [antonbabenko/pre-commit-terraform#421]
The example given for running the validate hook against non root modules advises to create a test root module and to exclude the top level directory as follows:

````
- id: terraform_validate
  exclude: '[^/]+$'
`````

This does not work in my ubuntu environment. Looking into the code I see it is using grep -e, however that is environment dependent. So for eample

`git ls-files | sort | grep -e "" | grep -v -e "[^/]+$"`
returns all files

adding the required escape
`git ls-files | sort | grep -e "" | grep -v -e "[^/]\+$"`
returns no files because all files have a non slash character at the end

correct pattern is to anchor to the start of the string
`git ls-files | sort | grep -e "" | grep -v -e "^[^/]\+$"`
returns only the files in a sub directory

There is an option to ue -E on grep so this works too
`git ls-files | sort | grep -e "" | grep -v -E "^[^/]+$"`
however that would be a change to pre-commit

All that said, I still can't get exclude to work from within the pre commit yaml - maybe some further escaping at play that I haven't got my head around yet.

What does work is to turn it around and use files instead of exclude
````
- id: terraform_validate
  files: tests/validate
`````

Maybe a note in the documentation and/or perhaps show using files instead of exclude would help other people.
Thanks.

**Comment by yermulnik** [comment]
> This does not work in my ubuntu environment. Looking into the code I see it is using grep -e, however that is environment dependent.

Yep, `grep` version and flavour may differ from OS to OS =(

Sorry I don't get the gist: what do you expect to be output once `exclude` is applied?
Please show the example of unfiltered files list and the one you'd expect to see once filter is applied.

**Comment by Noel-Jones** [comment]
Without the fix, which for me is to use files to only include the test directory, you get this, indicating that terraform validate is run against the root of the module (which is not a terraform root module). As in
````
Terraform validate.......................................................Failed
- hook id: terraform_validate
- exit code: 1

Validation failed: .
╷
│ Error: missing provider provider["registry.terraform.io/hashicorp/azurerm"].core
````
Sorry if it wasn't clear but the grep commands were from the pre-commit source code. So what needs to be returned from the grep is just the test directory files, e.g.

````
~/reps/Modules/storage-new$ git ls-files | sort | grep -e "" | grep -v -e "^[^/]\+$"
tests/validate/.terraform.lock.hcl
tests/validate/main.tf
````

But what is actually returned using the documentd exclude is this
````
~/reps/Modules/storage-new$ git ls-files | sort | grep -e "" | grep -v -e "[^/]+$"
.commitlintrc.yml
.gitignore
.pre-commit-config.yaml
.releaserc.yml
.terraform.lock.hcl
CHANGELOG.md
README.md
azure-pipelines.yml
main.tf
outputs.tf
tests/validate/.terraform.lock.hcl
tests/validate/main.tf
variables.tf
versions.tf
````

So pre-commit then goes on to run against every directory, the exclude is not having any impact.

**Comment by yermulnik** [comment]
So GNU grep by default uses Basic RE and hence what you're looking for is: `^[^/]\+$`
Snippet from `man 1 grep`:
```
> man grep | sed -n '/Basic vs Extended Regular Expressions/,/^$/p'
   Basic vs Extended Regular Expressions
       In  basic regular expressions the meta-characters ?, +, {, |, (, and ) lose their special meaning; instead
       use the backslashed versions \?, \+, \{, \|, \(, and \).
```
Please feel free to contribute by creating PR to README. Thanks.
@MaxymVlasov Could you please check whether `grep` on MacOS by default uses Basic RE or Extended RE?

**Comment by Noel-Jones** [comment]
Well I guess that was a case of TLDR? Go back to the first 4 lines.

The example given for running the validate hook against non root modules advises to create a test root module and to exclude the top level directory as follows:

````
- id: terraform_validate
  exclude: '[^/]+$'
````

**This does not work in my ubuntu environment**.

**Comment by yermulnik** [comment]
@Noel-Jones Yep, I got that and that's why I suggested you to help us improving this tool as you're familiar with the use case («_Please feel free to contribute by creating PR to README_»). if Max confirms the `'[^/]+$'` works as expected on MacOS, then would you mind adding a note for GNU `grep` users that they need to use `'^[^/]\+$'` instead? Appreciate yuour contribution in advance 🤝

**Comment by yermulnik** [comment]
@Noel-Jones Should I keep this issue opened until then?

**Comment by Noel-Jones** [comment]
Sorry to say that to eat humble pie. I don't know what version of pre-commit I was looking at or testing before I raised this. Suffice to say forget all that stuff about grep. The current version of pre-commit (everything upgraded now so 2.20.0) uses python regex.

The issue still stands though. With the exclude as documented ALL files are excluded:

````
     - id: terraform_validate
      exclude: '[^/]+$'
````

output:
`Terraform validate...................................(no files to check)Skipped'

This is  because all files will have a none slash character at the end of the file path. Wth the regex correctly anchored to check the whole file path:

````
    - id: terraform_validate
      exclude: '^[^/]+$'
````

output:
`Terraform validate.......................................................Passed'

And I can see that the .terraform directory is created in the subdirectory only, as expected.

**Comment by MaxymVlasov** [comment]
@yermulnik how I can confirm something about MacOS (except googling) if I never use anything Apple related (except one case when I ask Mac for a week to understand why Mac wouldn't work with Makefile as expected)? :) 

If you need someone with Mac - ask Anton

**Comment by MaxymVlasov** [comment]
By the way, where in the code `grep` mentioned? It uses `re.compile`

https://github.com/pre-commit/pre-commit/blob/9c55cc3954ecdacc288fba5deaa67d65bbd5972b/pre_commit/commands/run.py#L59-L69

**Comment by yermulnik** [comment]
> I never use anything Apple related

Oh mate, my apologies, I somehow thought you're on Mac (I don't know why I was under this impression though 🤷🏻). Hope it's not a deadly offense though 🤣

**Comment by yermulnik** [comment]
> The issue still stands though. With the exclude as documented ALL files are excluded:

@Noel-Jones So the README still needs an amendment to address this, right? Are you open to contribute?

**Comment by MaxymVlasov** [comment]
Looks like there is only one place where that in README is present. 
#422

**Comment by antonbabenko** [comment]
This issue has been resolved in version 1.74.2 :tada:
